mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java

```diff
@@ -19,21 +19,445 @@
+              .schema(schema)
+              .createWriterFunc(GenericParquetWriter::buildWriter)
+              .named(fileFormat.name())
+              .build();
+          break;
+
+        case ORC:
+          this.appender = ORC.write(dataFile)
+              .schema(schema)
+              .createWriterFunc(GenericOrcWriter::buildWriter)
+              .build();
+          break;
+
+        default:
+          throw new UnsupportedOperationException("Cannot write format: " + fileFormat);
+      }
+      LOG.info("IcebergRecordWriter is created in {} with {}", location, fileFormat);
+    }
+
+    @Override
+    public void write(Writable row) {
+      Preconditions.checkArgument(row instanceof IcebergWritable);
+
+      // TODO partition handling
+      appender.add(((IcebergWritable) row).record());
+    }
+
+    @Override
+    public void write(NullWritable key, IcebergWritable value) {
+      write(value);
+    }
+
+    @Override
+    public void close(boolean abort) throws IOException {
+      appender.close();
+      if (!abort) {
+        fileData.put(taskAttemptId, new ClosedFileData(location, fileFormat, appender.length(), appender.metrics()));
+      }
+    }
+
+    @Override
+    public void close(org.apache.hadoop.mapreduce.TaskAttemptContext context) throws IOException {
+      close(false);
+    }
+
+    @Override
+    public void close(Reporter reporter) throws IOException {
+      close(false);
+    }
+  }
+
+  /**
+   * A dummy committer - not related to the Hive transactions.
+   */
+  public static final class IcebergOutputCommitter extends OutputCommitter {
+
+    @Override
+    public void setupJob(JobContext jobContext) {
+      // do nothing.
+    }
+
+    @Override
+    public void setupTask(TaskAttemptContext taskAttemptContext) {
+      // do nothing.
+    }
+
+    @Override
+    public boolean needsTaskCommit(TaskAttemptContext taskAttemptContext) {
+      return true;
+    }
+
+    @Override
+    public void commitTask(TaskAttemptContext context) throws IOException {
+      ClosedFileData closedFileData = fileData.remove(context.getTaskAttemptID());
+      String commitFileLocation = generateCommitFileLocation(context.getJobConf(), context.getTaskAttemptID());
+
+      // If we created a new data file, then create the committed file for this
+      if (closedFileData != null) {
+        createCommittedFileFor(new HadoopFileIO(context.getJobConf()), closedFileData, commitFileLocation);
+      } else {
+        TaskType taskType = context.getTaskAttemptID().getTaskID().getTaskType();
+        boolean isWrite = context.getJobConf().getBoolean(HiveIcebergStorageHandler.WRITE_KEY, false);
+        boolean mapOnly = context.getJobConf().getNumReduceTasks() == 0;
+
+        // If we writing and the task is either reducer or a map in a map-only job then write an empty commit file
+        if (isWrite && (TaskType.REDUCE.equals(taskType) || (TaskType.MAP.equals(taskType) && mapOnly))) {
```

Comment:
I'm not sure I understand why these need to be checked. Is `commitTask` called when not writing?

Reply:
The JobRunner will call OutputCommitter.needsTaskCommit() to check if the OutputCommitter.commitTask() / OutputCommitter.abortTask() call is needed. If needsTaskCommit() returns true then the commitTask() will be called for the Mapper or the Reducer independently.
I have tested this with the following queries:
```
insert into purchases values (25,"3","2020-09-03"),(56,4,"1900-01-01");
insert into purchases select * from purchases;
insert into purchases select * from purchases order by id;
```
Query1: translated to a map only job directly writing to the table in the mapper
Query2: reads, and writes the same table in a map only job
Query3: reads the table in the mappers, and writes the table in the reducers

Your comment made me think about how to write a more self-explaining  code, and put the checks to the needsTaskCommit(). This might make this easier to understand (the downside is that the decision is separated from the actual execution)

Reply:
Doesn't every task need a commit if we need to produce this file? Then we can separate the logic, but `needsTaskCommit` should just return `true`.

Reply:
FileSinkOperators are writing data. These are the last operators in a job.
So we should write the commit file at the end of the Reducers or at the end of the Mappers for the map only jobs. We do not need to generate the commit file at the end of the Map tasks if a Reduce task will follow, as only the Reducer will write the data.

Or alternatively we can always write the file and deal with multiple empty files. This would work too, and less reliant on assumptions around the place of the FileSinkOperators in the Hive workflow,  but definitely less performant. That is why I went for the solution where we decide on the file handling in the needsTaskCommit, and not committing every task.
Would you still go for the more general approach, with more empty commit files?